Quil program to Cirq circuit conversion

karalekas opened the pull request:
```python
In [1]: from cirq.contrib.quil_import import circuit_from_quil

In [2]: circuit_from_quil("""
   ...: I 0
   ...: I 1
   ...: I 2
   ...: X 0
   ...: Y 1
   ...: Z 2
   ...: H 0
   ...: S 1
   ...: T 2
   ...: PHASE(pi/8) 0
   ...: PHASE(pi/8) 1
   ...: PHASE(pi/8) 2
   ...: RX(pi/2) 0
   ...: RY(pi/2) 1
   ...: RZ(pi/2) 2
   ...: CZ 0 1
   ...: CNOT 1 2
   ...: CPHASE(pi/2) 0 1
   ...: SWAP 1 2
   ...: ISWAP 0 1
   ...: XY(pi/2) 1 2
   ...: CCNOT 0 1 2
   ...: CSWAP 0 1 2
   ...: MEASURE 0 ro[0]
   ...: MEASURE 1 ro[1]
   ...: MEASURE 2 ro[2]
   ...: """)
Out[2]:
0: ───I───X───H───Z^(1/8)───Rx(0.5π)───@───────@───────────iSwap───────────────@───@───M('ro[0]')───
                                       │       │           │                   │   │
1: ───I───Y───S───Z^(1/8)───Ry(0.5π)───@───@───@^0.5───×───iSwap───iSwap───────@───×───M('ro[1]')───
                                           │           │           │           │   │
2: ───I───Z───T───Z^(1/8)───Rz(0.5π)───────X───────────×───────────iSwap^0.5───X───×───M('ro[2]')───
```

karalekas committed:
First attempt at Quil program to Cirq circuit conversion

karalekas committed:
Add .python-version to .gitignore

karalekas committed:
Add pyquil to contrib-requirements.txt

karalekas committed:
Bugfix: Quil export of common gates off by a factor of pi

karalekas committed:
Second pass: defgates, all standard gates, raise errors

karalekas committed:
Make check/format-incremental happy (was using black)

karalekas committed:
Add pyquil to stubless list in mypy.ini (for check/mypy)

karalekas committed:
Missed some quil conversion fixes in non-common gates

karalekas committed:
Update the quil output tests to match added factor of pi

karalekas committed:
Finish full code coverage of quil_import module

karalekas committed:
Make docstrings googley-er and add quil_import to api.rst

rmlarose commented on cirq/contrib/quil_import/quil_test.py:
```diff
@@ -0,0 +1,185 @@
+    with pytest.raises(UnsupportedQuilGate):
+        circuit_from_quil("FREDKIN 0 1 2")
+
+    with pytest.raises(UnsupportedQuilGate):
+        circuit_from_quil("TOFFOLI 0 1 2")
```
Why isn't Toffoli supported? `cirq.ops.TOFFOLI` (and also `cirq.ops.CCNOT`).

karalekas replied:
There is no TOFFOLI gate in Quil

rmlarose replied:
Ah ok, so TOFOLLI 0 1 2 isn't a valid Quil program. Does it still make sense to test then?

willzeng replied:
Would it be alright to convert it to a `CCNOT` gate for pyQuil?

karalekas replied:
This is helpful feedback -- clearly `UnsupportedQuilGate` is a bad name for this error. I added this test to get 100% coverage (I needed to cover the branch that hits this error) and the point of the error is to say "you've provided a gate that is not supported by Quil and also is not defined via a defgate." Thus, `UndefinedQuilGate` is probably a better name.

rmlarose commented on cirq/contrib/quil_import/quil_test.py:
```diff
@@ -0,0 +1,185 @@
+
+
+def test_unsupported_quil_gate():
+    with pytest.raises(UnsupportedQuilGate):
+        circuit_from_quil("FREDKIN 0 1 2")
```
Why isn't Fredkin supported? `cirq.ops.FREDKIN`

karalekas replied:
ditto ^

rmlarose commented on cirq/circuits/quil_output_test.py:
```diff
@@ -36,9 +36,9 @@ def test_single_gate_no_parameter():
+    assert (str(output) == f"""# Created using Cirq.
 
-RX(0.5) 0\n""")
+RX({np.pi/2}) 0\n""")
```
```suggestion
RX({np.pi / 2}) 0\n""")
```

rmlarose replied:
Here and elsewhere. Unless there's some reason to use different formatting in f strings?

rmlarose commented on cirq/contrib/quil_import/quil.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def cphase(param: float) -> CZPowGate:
+    """The angle parameter of pyQuil's CPHASE gate and the exponent of
+    Cirq's CZPowGate differ by a factor of pi.
```
```suggestion
    """Returns a controlled-phase gate as a Cirq CZPowGate with the
    input param.
    
    The angle parameter of pyQuil's CPHASE gate and the exponent of
    Cirq's CZPowGate differ by a factor of pi.
```

rmlarose replied:
I could maybe see docstrings asking to be changed. To me they are fine as is.

rmlarose commented on cirq/contrib/quil_import/quil.py:
```diff
@@ -0,0 +1,269 @@
+    Returns:
+        A MatrixGate equivalent to a CPHASE01 gate of given angle.
+    """
+    cphase01_matrix = np.array([[1, 0, 0, 0], [0, np.exp(1j * phi), 0, 0],
+                                [0, 0, 1, 0], [0, 0, 0, 1]])
```
```suggestion
    cphase01_matrix = np.array([
        [1, 0, 0, 0], 
        [0, np.exp(1j * phi), 0, 0],
        [0, 0, 1, 0],
        [0, 0, 0, 1]
    ])
```

rmlarose replied:
Here and elsewhere. Maybe be explicit about the data type as well?

karalekas replied:
Made this change and `check/format-incremental` complains 🤷 -- going to undo

rmlarose commented on cirq/contrib/quil_import/quil.py:
```diff
@@ -0,0 +1,269 @@
+    Returns:
+        A MatrixGate equivalent to a CPHASE10 gate of given angle.
+    """
+    cphase10_matrix = np.array([[1, 0, 0, 0], [0, 1, 0, 0],
+                                [0, 0, np.exp(1j * phi), 0], [0, 0, 0, 1]])
```
Formatting.

rmlarose commented on cirq/contrib/quil_import/quil.py:
```diff
@@ -0,0 +1,269 @@
+            quil_gate_name = inst.name
+            quil_gate_params = inst.params
+            line_qubits = list(LineQubit(q.index) for q in inst.qubits)
+            defgates_and_supported_gates = dict(**defgates, **SUPPORTED_GATES)
```
Since `SUPPORTED_GATES` is constant and only `defgates` is changing through the loop, maybe

```python
defgates_and_supported_gates.update(defgates)
```

instead?

rmlarose commented on cirq/contrib/quil_import/quil.py:
```diff
@@ -0,0 +1,269 @@
+            quil_gate_params = inst.params
+            line_qubits = list(LineQubit(q.index) for q in inst.qubits)
+            defgates_and_supported_gates = dict(**defgates, **SUPPORTED_GATES)
+            if quil_gate_name not in defgates_and_supported_gates:
```
Is it possible for a DEFGATE to appear after its declaration in Quil? In other words, does the gate have to be defined before it is used in a Quil program? Or could the definition appear anywhere in the Quil program?

karalekas replied:
it has to be defined beforehand afaik

rmlarose reviewed (commented):
Great stuff @karalekas, see inline comments.

Maybe a good idea to split this into two PRs? One for fixing the current Quil output and one for adding the (py)Quil input.

karalekas committed:
Consistent formatting in matrices, add complex dtype

karalekas committed:
Rename UnsupportedQuilGate to UndefinedQuilGate

karalekas committed:
Be even more explicit in quil.py docstrings

karalekas committed:
Initialized defined_gates with SUPPORTED_GATES to simplify

karalekas committed:
Add space around binary ops in quil_output_test.py

karalekas committed:
Undo matrix formatting to appease check/format-incremental

karalekas committed:
Add pyquil to pip-list-dev-tools.txt

karalekas committed:
Add equivalent unitaries test to quil_output_test.py

karalekas committed:
Return Quil XY gate when ISwapPowGate has non-unity exponent

karalekas committed:
Small formatting changes to quil_output_test.py

karalekas commented:
@rmlarose Addressed all feedback, and verified that `check/all` works. Additionally, added a program unitary test to `quil_output_test.py`, and changed the quil output to `XY` for non-unity `ISwapPowGate`s. This PR has gotten big enough that I agree with your sentiment re: splitting it up into two. If you're happy with the current state I'll move forward on that.
